<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка-данных" data-toc-modified-id="Подготовка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка данных</a></span><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Предобработка данных</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Исследование-задачи" data-toc-modified-id="Исследование-задачи-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследование задачи</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Случайный лес</a></span></li><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Борьба-с-дисбалансом" data-toc-modified-id="Борьба-с-дисбалансом-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Борьба с дисбалансом</a></span><ul class="toc-item"><li><span><a href="#Увеличение-выборки" data-toc-modified-id="Увеличение-выборки-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Увеличение выборки</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений-на-увеличенной-выборке" data-toc-modified-id="Дерево-решений-на-увеличенной-выборке-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Дерево решений на увеличенной выборке</a></span></li><li><span><a href="#Случайный-лес-на-увеличенной-выборке" data-toc-modified-id="Случайный-лес-на-увеличенной-выборке-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Случайный лес на увеличенной выборке</a></span></li><li><span><a href="#Логистическая-регрессия-на-увеличенной-выборке" data-toc-modified-id="Логистическая-регрессия-на-увеличенной-выборке-3.1.3"><span class="toc-item-num">3.1.3&nbsp;&nbsp;</span>Логистическая регрессия на увеличенной выборке</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.1.4"><span class="toc-item-num">3.1.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Уменьшение-выборки" data-toc-modified-id="Уменьшение-выборки-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Уменьшение выборки</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений-на-уменьшенной-выборке" data-toc-modified-id="Дерево-решений-на-уменьшенной-выборке-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Дерево решений на уменьшенной выборке</a></span></li><li><span><a href="#Случайный-лес-на-уменьшенной-выборке" data-toc-modified-id="Случайный-лес-на-уменьшенной-выборке-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Случайный лес на уменьшенной выборке</a></span></li><li><span><a href="#Логистическая-регрессия-на-уменьшенной-выборке" data-toc-modified-id="Логистическая-регрессия-на-уменьшенной-выборке-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Логистическая регрессия на уменьшенной выборке</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.2.4"><span class="toc-item-num">3.2.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Комбинирование-увеличения-и-уменьшения-выборки" data-toc-modified-id="Комбинирование-увеличения-и-уменьшения-выборки-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Комбинирование увеличения и уменьшения выборки</a></span><ul class="toc-item"><li><span><a href="#Дерево-решений-на-комбинированной-выборке" data-toc-modified-id="Дерево-решений-на-комбинированной-выборке-3.3.1"><span class="toc-item-num">3.3.1&nbsp;&nbsp;</span>Дерево решений на комбинированной выборке</a></span></li><li><span><a href="#Случайный-лес-на-комбинированной-выборке" data-toc-modified-id="Случайный-лес-на-комбинированной-выборке-3.3.2"><span class="toc-item-num">3.3.2&nbsp;&nbsp;</span>Случайный лес на комбинированной выборке</a></span></li><li><span><a href="#Логистическая-регрессия-на-комбинированной-выборке" data-toc-modified-id="Логистическая-регрессия-на-комбинированной-выборке-3.3.3"><span class="toc-item-num">3.3.3&nbsp;&nbsp;</span>Логистическая регрессия на комбинированной выборке</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.3.4"><span class="toc-item-num">3.3.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Тестирование-модели" data-toc-modified-id="Тестирование-модели-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Тестирование модели</a></span><ul class="toc-item"><li><span><a href="#Случайный-лес,-модель-увеличенной-выборки" data-toc-modified-id="Случайный-лес,-модель-увеличенной-выборки-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Случайный лес, модель увеличенной выборки</a></span></li><li><span><a href="#Случайный-лес,-модель-уменьшенной-выборки" data-toc-modified-id="Случайный-лес,-модель-уменьшенной-выборки-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Случайный лес, модель уменьшенной выборки</a></span></li><li><span><a href="#Случайный-лес,-модель-комбинированной-выборки" data-toc-modified-id="Случайный-лес,-модель-комбинированной-выборки-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Случайный лес, модель комбинированной выборки</a></span></li><li><span><a href="#Проверка-на-адекватность" data-toc-modified-id="Проверка-на-адекватность-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Проверка на адекватность</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li></ul></div>

# Отток клиентов

**Описание проекта**

Из «Бета-Банка» стали уходить клиенты. Каждый месяц. Немного, но заметно. Банковские маркетологи посчитали: сохранять текущих клиентов дешевле, чем привлекать новых.

Нужно спрогнозировать, уйдёт клиент из банка в ближайшее время или нет. Вам предоставлены исторические данные о поведении клиентов и расторжении договоров с банком. 

**Цель проекта**

Постройте модель с предельно большим значением *F1*-меры. Чтобы сдать проект успешно, нужно довести метрику до 0.59. Проверьте *F1*-меру на тестовой выборке самостоятельно.

Дополнительно измеряйте *AUC-ROC*, сравнивайте её значение с *F1*-мерой.

**Описание данных**

`/datasets/Churn.csv` 

*Признаки:*

- 'RowNumber' — индекс строки в данных
- 'CustomerId' — уникальный идентификатор клиента
- 'Surname' — фамилия
- 'CreditScore' — кредитный рейтинг
- 'Geography' — страна проживания
- 'Gender' — пол
- 'Age' — возраст
- 'Tenure' — сколько лет человек является клиентом банка
- 'Balance' — баланс на счёте
- 'NumOfProducts' — количество продуктов банка, используемых клиентом
- 'HasCrCard' — наличие кредитной карты
- 'IsActiveMember' — активность клиента
- 'EstimatedSalary' — предполагаемая зарплата

*Целевой признак:*

- 'Exited' — факт ухода клиента

Источник данных: [https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling](https://www.kaggle.com/barelydedicated/bank-customer-churn-modeling)

In [158]:
# импорт библиотек

import pandas as pd

import numpy as np

from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression 
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler 
from sklearn.utils import shuffle
from sklearn.metrics import f1_score
from sklearn.metrics import roc_auc_score 
from sklearn.metrics import precision_score, recall_score

In [2]:
# отключу предупреждения
import warnings
warnings.filterwarnings('ignore')

## Подготовка данных

### Обзор данных

In [3]:
try:
    data = pd.read_csv('datasets/Churn.csv')
except FileNotFoundError as e:
    print(repr(e))
    data = pd.read_csv('/datasets/Churn.csv')

FileNotFoundError(2, 'No such file or directory')


In [4]:
# получаю общую сводку
data.info()

# проверю на корректный вывод
display(data.sample(5))

# проверяю имена колонок на пробелы
display(data.columns)

# проверю на явные дубликаты
print(f'Количество явных дубликатов: {data.duplicated().sum()}')

# описание
data.describe().T

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Surname          10000 non-null  object 
 3   CreditScore      10000 non-null  int64  
 4   Geography        10000 non-null  object 
 5   Gender           10000 non-null  object 
 6   Age              10000 non-null  int64  
 7   Tenure           9091 non-null   float64
 8   Balance          10000 non-null  float64
 9   NumOfProducts    10000 non-null  int64  
 10  HasCrCard        10000 non-null  int64  
 11  IsActiveMember   10000 non-null  int64  
 12  EstimatedSalary  10000 non-null  float64
 13  Exited           10000 non-null  int64  
dtypes: float64(3), int64(8), object(3)
memory usage: 1.1+ MB


RowNumber  CustomerId    Surname  CreditScore Geography  Gender  Age  \
7163       7164    15649379   Somayina          850    France  Female   46   
5065       5066    15775709      Nucci          832    France  Female   27   
9771       9772    15771323  Panicucci          480     Spain    Male   39   
8859       8860    15743456     Birnie          715    France  Female   32   
9605       9606    15733444   Phillips          736    France  Female   29   

      Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
7163     3.0       0.00              2          1               1   
5065    10.0   98590.25              1          1               0   
9771     5.0  121626.90              1          1               1   
8859    10.0       0.00              2          1               0   
9605     9.0       0.00              2          0               0   

      EstimatedSalary  Exited  
7163        187980.21       0  
5065         30912.89       0  
9771         82438.13       0  
8859         60907.49       0  
9605        176152.70       0

Index(['RowNumber', 'CustomerId', 'Surname', 'CreditScore', 'Geography',
       'Gender', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'HasCrCard',
       'IsActiveMember', 'EstimatedSalary', 'Exited'],
      dtype='object')

Количество явных дубликатов: 0


count          mean           std          min  \
RowNumber        10000.0  5.000500e+03   2886.895680         1.00   
CustomerId       10000.0  1.569094e+07  71936.186123  15565701.00   
CreditScore      10000.0  6.505288e+02     96.653299       350.00   
Age              10000.0  3.892180e+01     10.487806        18.00   
Tenure            9091.0  4.997690e+00      2.894723         0.00   
Balance          10000.0  7.648589e+04  62397.405202         0.00   
NumOfProducts    10000.0  1.530200e+00      0.581654         1.00   
HasCrCard        10000.0  7.055000e-01      0.455840         0.00   
IsActiveMember   10000.0  5.151000e-01      0.499797         0.00   
EstimatedSalary  10000.0  1.000902e+05  57510.492818        11.58   
Exited           10000.0  2.037000e-01      0.402769         0.00   

                         25%           50%           75%          max  
RowNumber            2500.75  5.000500e+03  7.500250e+03     10000.00  
CustomerId       15628528.25  1.569074e+07  1.575323e+07  15815690.00  
CreditScore           584.00  6.520000e+02  7.180000e+02       850.00  
Age                    32.00  3.700000e+01  4.400000e+01        92.00  
Tenure                  2.00  5.000000e+00  7.000000e+00        10.00  
Balance                 0.00  9.719854e+04  1.276442e+05    250898.09  
NumOfProducts           1.00  1.000000e+00  2.000000e+00         4.00  
HasCrCard               0.00  1.000000e+00  1.000000e+00         1.00  
IsActiveMember          0.00  1.000000e+00  1.000000e+00         1.00  
EstimatedSalary     51002.11  1.001939e+05  1.493882e+05    199992.48  
Exited                  0.00  0.000000e+00  0.000000e+00         1.00

In [5]:
'Доля пропусков в признаке Tenure от общего количества строк:', data.Tenure.isna().mean()

('Доля пропусков в признаке Tenure от общего количества строк:', 0.0909)

In [6]:
print('Количество классов в целевом признаке: \n')
print(data.Exited.value_counts())

print('\n ////////////////////////// \n')

print('В процентном соотношении:')
data.Exited.value_counts().div(
    data.Exited.value_counts().sum()
).mul(100).round(2)

Количество классов в целевом признаке: 

0    7963
1    2037
Name: Exited, dtype: int64

 ////////////////////////// 

В процентном соотношении:


0    79.63
1    20.37
Name: Exited, dtype: float64

Предварительные замечания:
- Признаки ['RowNumber', 'CustomerId', 'Surname'] не требуются для обучения модели;
- Названия признаков следует привести к нижнему регистру, для удобства работы;
- Пропуски в признаке ['Tenure'], 9.09% от общего количества строк;
- Тип данных ['Tenure'] - float. Можно исопльзовать integer;
- Дисбаланс классов, соотношение приближено к 4:1.

### Предобработка данных

Проведу базовую очистку данных на основе предварительных замечаний.

In [7]:
# избавлюсь от первых трех колонок, приведу оставшиеся к нижнему регистру
data.drop(data.columns[[0,1,2]], axis=1, inplace=True)
data.columns = data.columns.str.lower()

# проверю результат
data.columns

Index(['creditscore', 'geography', 'gender', 'age', 'tenure', 'balance',
       'numofproducts', 'hascrcard', 'isactivemember', 'estimatedsalary',
       'exited'],
      dtype='object')

In [8]:
# прставлю разделители
data.columns = ['credit_score', 'geography',
       'gender', 'age', 'tenure', 'balance', 'num_of_products', 'has_crcard',
       'is_active_member', 'estimated_salary', 'exited']

In [9]:
# так как не к чему привязать значения tenure - просто заполню пропуски средним
data.tenure.fillna(data.tenure.mean().round(), inplace=True)

data.tenure.isna().sum()

0

In [10]:
# поменяю тип данных признака tenure
data['tenure'] = data['tenure'].astype('int32')

Применю техники преобразования данных в форму подходящую алгоритмам машинного обучения. 

Так как я планирую исопльзовать модели Дерева Решений и Случайного леса, то для обработк категориальных признаков использую технику прямого кодирования (One-hot Encoding).

In [11]:
# использую drop_first=True чтобы избежать дамми-ловушки
data_ohe = pd.get_dummies(data, drop_first=True)
target = data_ohe['exited']
features = data_ohe.drop('exited', axis=1)

In [12]:
features.sample(10)

credit_score  age  tenure    balance  num_of_products  has_crcard  \
8731           727   41      10       0.00                2           0   
1964           800   40       5   97764.41                1           1   
5727           535   38       8  127475.24                1           0   
1667           606   46       5       0.00                2           1   
8313           639   38       5       0.00                1           1   
2566           660   29       6  180520.29                1           1   
8838           558   33       0  108477.49                1           1   
1530           611   26       2  107508.93                2           1   
2611           566   42       3  108010.78                1           1   
9679           472   19       9       0.00                2           1   

      is_active_member  estimated_salary  geography_Germany  geography_Spain  \
8731                 1          47468.56                  0                1   
1964                 0          98640.15                  0                0   
5727                 0          60775.76                  1                0   
1667                 1         183717.94                  0                1   
8313                 0          42862.82                  0                0   
2566                 1         123850.58                  1                0   
8838                 1         109096.71                  0                0   
1530                 1         120801.65                  0                0   
2611                 1         157486.10                  0                0   
9679                 0           3453.40                  0                0   

      gender_Male  
8731            0  
1964            0  
5727            1  
1667            0  
8313            1  
2566            0  
8838            1  
1530            0  
2611            1  
9679            1

Разобью данные да обучающую, валидационную и тестовую выборки в соотношении 6:2:2.

In [15]:
# для фиксации результата использую число 568702396, полученное генератором случайных чисел
features_train, features_valid, target_train, target_valid = train_test_split(
    features,
    target,
    test_size=0.4,
    random_state=568702396
)

In [16]:
# воспользуюсь тем же методом для разделения валидационной и тестовой выборок
features_test, features_valid, target_test, target_valid = train_test_split(
    features_valid,
    target_valid,
    test_size=0.5,
    random_state=568702396
)

Масштаб признаков значительно различается. Использую метод стандартизации данных.

In [17]:
# выделю признаки нуждающиеся в стандартизации
numeric = ['credit_score', 'age', 'balance', 'estimated_salary']

# обучу метод на тренировочный выборке
scaler = StandardScaler()
scaler.fit(features_train[numeric]) 

# применю метод к тренировочной и тестовой выборкам
features_train[numeric] = scaler.transform(features_train[numeric])
features_valid[numeric] = scaler.transform(features_valid[numeric])
features_test[numeric] = scaler.transform(features_test[numeric])

In [18]:
features_train.sample(10)

credit_score       age  tenure   balance  num_of_products  has_crcard  \
2589      1.327863  0.462217       8  0.755387                1           1   
9513     -1.396872  0.841018       5  0.844233                1           0   
5592     -0.764899 -1.905292       6 -1.236677                2           0   
4818      1.058497 -0.295386       1  0.159913                1           1   
5330     -0.350491 -0.958289       9 -1.236677                1           1   
5251     -0.319410 -0.011285       5 -1.236677                2           1   
2557      1.939115 -1.242390       4  0.209163                2           1   
9883      0.291842  0.178116       2  1.144982                1           1   
4289      0.840933 -0.200686       1  0.836907                1           1   
3801      0.509406 -0.768888       6  0.604693                1           0   

      is_active_member  estimated_salary  geography_Germany  geography_Spain  \
2589                 0          0.760018                  0                0   
9513                 1          0.819932                  0                0   
5592                 0         -0.489678                  0                1   
4818                 1          0.081159                  0                0   
5330                 1          1.149344                  0                0   
5251                 1          1.001507                  0                0   
2557                 0          1.296363                  0                0   
9883                 0         -1.497602                  0                0   
4289                 1          1.386997                  0                1   
3801                 0          0.654442                  0                1   

      gender_Male  
2589            1  
9513            1  
5592            1  
4818            0  
5330            1  
5251            1  
2557            1  
9883            1  
4289            0  
3801            1

### Вывод

По итогам предобработки:
- Отброшены не требующиеся для обучения признаки ['RowNumber', 'CustomerId', 'Surname'];
- Названия столбцов приведены к единому виду;
- Пропущенные значения признака tenure заполнены средним значением по признаку;
- Выбрана техника OHE, категориальыне признаки приведены к виду пригодному для алгоритмов ML;
- Получены обучающая, валидационная и тестовая выборки;
- Стандартизированы количественные признаки с выделяющимся масштабом. 

Подготовка данных завершена. Проблема дисбаланса будет исследована после построения модели на исходных данных.

## Исследование задачи

Цель - простроение модели на основании исторических данных банка, способной спрогнозировать уход клиента на основании его данных.

В задании дано требование достижения моделью минимального значения F1-меры равной или более 0.59.
Допонительно требуется измерять AUC-ROC с целью сравнения с F1-мерой.

Для решения задачи воспользуюсь моделями Дерева Решений, Случайного Леса и Логистической Регрессии.

Первые модели построю без учета дисбаланса классов. Затем применю методы устранения дисбаланса и буду работать над улучшением моделей до приемлемых показателей.

Для данного проекта не стану использовать метод GridSearch CV, соберу простые циклы.

### Дерево решений

In [56]:
best_criteria = 0
best_depth = 0
best_split = 0
best_leaf = 0
best_f1 = 0

criteria = ['gini', 'entropy']

for i in range(len(criteria)):
    for depth in range(2, 6):
        for split in range(2, 6):
            for leaf in range(2, 6):
                model_tree = DecisionTreeClassifier(
                    random_state=568702396,
                    criterion=criteria[i],
                    max_depth=depth,
                    min_samples_split=split,
                    min_samples_leaf=leaf)
                
                model_tree.fit(features_train, target_train)
                predictions_valid = model_tree.predict(features_valid)
                
                f1 = f1_score(target_valid, predictions_valid)
                if f1 > best_f1:
                    best_f1 = f1
                    best_criteria = criteria[i]
                    best_depth = depth
                    best_split = split
                    best_leaf = leaf
                    
                
print(f'Максимальное достигнутое значение f1-меры: {best_f1}, при параметрах:')
print(f'- критерион - {best_criteria};')
print(f'- глубина дерева решений = {best_depth};')
print(f'- минимальное количество примеров для разделения: {best_split};')
print(f'- минимальное количество выборок для листового узла: {best_leaf}.')

Максимальное достигнутое значение f1-меры: 0.5688073394495413, при параметрах:
- критерион - entropy;
- глубина дерева решений = 5;
- минимальное количество примеров для разделения: 2;
- минимальное количество выборок для листового узла: 3.


Сравню полученное значение f1-меры с показателем AUC-ROC.

In [45]:
probabilities_valid = model_tree.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

auc_roc

0.837880834529837

### Случайный лес

In [57]:
best_criteria = 0
best_depth = 0
best_est = 0
best_leaf = 0
best_f1 = 0

criteria = ['gini', 'entropy']

for i in range(len(criteria)):
    for est in range(2, 53, 5):
        for depth in range(2, 6):
            for leaf in range(2, 6):
               model_forest = RandomForestClassifier(
                   random_state=568702396,
                   criterion=criteria[i],
                   max_depth=depth,
                   n_estimators=est,
                   min_samples_leaf=leaf) 
               
               model_forest.fit(features_train, target_train)
               predictions_valid = model_forest.predict(features_valid)
               
               f1 = f1_score(target_valid, predictions_valid)
               if f1 > best_f1:
                   best_f1 = f1
                   best_criteria = criteria[i]
                   best_depth = depth
                   best_est = est
                   best_leaf = leaf

print(f'Максимальное достигнутое значение f1-меры: {best_f1} при модели:')
print(f'- количество деревьев: {best_est};')
print(f'- критерион - {best_criteria};')
print(f'- глубина дерева решений: {best_depth};')
print(f'- минимальное количество выборок для листового узла: {best_leaf}.')

Максимальное достигнутое значение f1-меры: 0.5718390804597702 при модели:
- количество деревьев: 2;
- критерион - entropy;
- глубина дерева решений: 5;
- минимальное количество выборок для листового узла: 5.


Сравню полученное значение f1-меры с показателем AUC-ROC.

In [44]:
probabilities_valid = model_forest.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

auc_roc

0.8462256605959847

### Логистическая регрессия

In [42]:
best_iter = 0
best_f1 = 0
best_solver = 0

solv = ['newton-cg', 'lbfgs', 'liblinear', 'saga']

for i in range(len(solv)):
    for itera in [25, 50, 100, 200, 400, 800]:
        model_regression = LogisticRegression(random_state=568702396, solver=solv[i], max_iter=itera)
        
        model_regression.fit(features_train, target_train)
        predictions_valid = model_regression.predict(features_valid)
        
        f1 = f1_score(target_valid, predictions_valid)
        if f1 > best_f1:
            best_f1 = f1
            best_iter = itera
            best_solver = solv[i]
 
print(f'Максимальное достигнутое значение f1-меры: {best_f1}, при параметрах:')
print(f'Лучшее количество итераций: {best_iter};')
print(f'Лучший метод: {best_solver}.')

Максимальное достигнутое значение f1-меры: 0.3473684210526316, при параметрах:
Лучшее количество итераций: 25;
Лучший метод: lbfgs.


Очень низкий показатель. Добавлю параметр class_weight='balanced'.

In [43]:
best_iter = 0
best_f1 = 0
best_solver = 0

solv = ['newton-cg', 'lbfgs', 'liblinear', 'saga']

for i in range(len(solv)):
    for itera in [25, 50, 100, 200, 400, 800]:
        model_regression = LogisticRegression(random_state=568702396, solver=solv[i], max_iter=itera, class_weight='balanced')
        model_regression.fit(features_train, target_train)
        predictions_valid = model_regression.predict(features_valid)
        f1 = f1_score(target_valid, predictions_valid)
        if f1 > best_f1:
            best_f1 = f1
            best_iter = itera
            best_solver = solv[i]
 
print(f'Максимальное достигнутое значение f1-меры: {best_f1}, при параметрах:')
print(f'Лучшее количество итераций: {best_iter};')
print(f'Лучший метод: {best_solver}.')

Максимальное достигнутое значение f1-меры: 0.4873646209386281, при параметрах:
Лучшее количество итераций: 25;
Лучший метод: newton-cg.


Сравню полученное значение f1-меры с показателем AUC-ROC.

In [47]:
probabilities_valid = model_regression.predict_proba(features_valid)
probabilities_one_valid = probabilities_valid[:, 1]

auc_roc = roc_auc_score(target_valid, probabilities_one_valid)

auc_roc

0.764691805838938

### Вывод

Используя исходные данные подвергнутые минимально-необходимой обработке, не удалось достигнуть желаемого показателя f1-меры. При этом для улучшения f1-меры логистической регрессии пришлось использовать параметр class_weight='balanced'.

Показатели AUC-ROC трех моделей значительно превышают показатели f1-меры. Это связано с тем что дисбаланс классов не оказывает значительного значения на AUC-ROC, также эта метрика растет в том числе от False Positive результатов.

Для улучшения моделей необходимо подобрать метод решения проблемы дисбаланса классов.

## Борьба с дисбалансом

### Увеличение выборки

Для увеличения выборки создам функциию, разделяющую выборку в соответствии со значением целевого признака, увеличивающую количество строк соотвутствующих положительному признаку и меняющую порядок строк результата.

In [80]:
def upsample(features, target, repeat):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]

    target_zeros = target[target == 0]
    target_ones = target[target == 1]
    
    features_upsampled = pd.concat([features_zeros] + [features_ones] * repeat)
    target_upsampled = pd.concat([target_zeros] + [target_ones] * repeat)
        
    features_upsampled, target_upsampled = shuffle(features_upsampled, target_upsampled, random_state=568702396)

    return features_upsampled, target_upsampled 

features_upsampled, target_upsampled = upsample(features_train, target_train, 2)

In [81]:
# сравню размеры
display(features_upsampled.shape)
target_upsampled.shape

(7234, 11)

(7234,)

In [82]:
# проверю баланс классов
print('Количество классов в целевом признаке: \n')
print(target_upsampled.value_counts())

print('\n ////////////////////////// \n')

print('В процентном соотношении:')
target_upsampled.value_counts().div(
    target_upsampled.value_counts().sum()
).mul(100).round(2)

Количество классов в целевом признаке: 

0    4766
1    2468
Name: exited, dtype: int64

 ////////////////////////// 

В процентном соотношении:


0    65.88
1    34.12
Name: exited, dtype: float64

Переобучу модели на увеличенной выборке, повторно подобрав параметры с помощью цикла.

#### Дерево решений на увеличенной выборке

In [116]:
best_criteria = 0
best_depth = 0
best_split = 0
best_leaf = 0
best_f1 = 0

criteria = ['gini', 'entropy']

for i in range(len(criteria)):
    for depth in range(2, 6):
        for split in range(2, 6):
            for leaf in range(2, 6):
                model_tree_up = DecisionTreeClassifier(
                    random_state=568702396,
                    criterion=criteria[i],
                    max_depth=depth,
                    min_samples_split=split,
                    min_samples_leaf=leaf)
                
                model_tree_up.fit(features_upsampled, target_upsampled)
                predictions_valid = model_tree_up.predict(features_valid)
                
                f1 = f1_score(target_valid, predictions_valid)
                if f1 > best_f1:
                    best_f1 = f1
                    best_criteria = criteria[i]
                    best_depth = depth
                    best_split = split
                    best_leaf = leaf
                    
                
print(f'Максимальное достигнутое значение f1-меры: {best_f1}, при параметрах:')
print(f'- критерион - {best_criteria};')
print(f'- глубина дерева решений = {best_depth};')
print(f'- минимальное количество примеров для разделения: {best_split};')
print(f'- минимальное количество выборок для листового узла: {best_leaf}.')

Максимальное достигнутое значение f1-меры: 0.5947955390334572, при параметрах:
- критерион - gini;
- глубина дерева решений = 5;
- минимальное количество примеров для разделения: 2;
- минимальное количество выборок для листового узла: 5.


#### Случайный лес на увеличенной выборке

In [117]:
best_criteria = 0
best_depth = 0
best_est = 0
best_leaf = 0
best_f1 = 0

criteria = ['gini', 'entropy']

for i in range(len(criteria)):
    for est in range(2, 53, 5):
        for depth in range(2, 6):
            for leaf in range(2, 6):
               model_forest_up = RandomForestClassifier(
                   random_state=568702396,
                   criterion=criteria[i],
                   max_depth=depth,
                   n_estimators=est,
                   min_samples_leaf=leaf) 
               
               model_forest_up.fit(features_upsampled, target_upsampled)
               predictions_valid = model_forest_up.predict(features_valid)
               
               f1 = f1_score(target_valid, predictions_valid)
               if f1 > best_f1:
                   best_f1 = f1
                   best_criteria = criteria[i]
                   best_depth = depth
                   best_est = est
                   best_leaf = leaf

print(f'Максимальное достигнутое значение f1-меры: {best_f1} при модели:')
print(f'- количество деревьев: {best_est};')
print(f'- критерион - {best_criteria};')
print(f'- глубина дерева решений: {best_depth};')
print(f'- минимальное количество выборок для листового узла: {best_leaf}.')

Максимальное достигнутое значение f1-меры: 0.6172506738544474 при модели:
- количество деревьев: 17;
- критерион - entropy;
- глубина дерева решений: 5;
- минимальное количество выборок для листового узла: 4.


#### Логистическая регрессия на увеличенной выборке

In [118]:
best_iter = 0
best_f1 = 0
best_solver = 0

solv = ['newton-cg', 'lbfgs', 'liblinear', 'saga']

for i in range(len(solv)):
    for itera in [25, 50, 100, 200, 400, 800]:
        model_regression_up = LogisticRegression(random_state=568702396, solver=solv[i], max_iter=itera)
        
        model_regression_up.fit(features_upsampled, target_upsampled)
        predictions_valid = model_regression_up.predict(features_valid)
        
        f1 = f1_score(target_valid, predictions_valid)
        if f1 > best_f1:
            best_f1 = f1
            best_iter = itera
            best_solver = solv[i]
 
print(f'Максимальное достигнутое значение f1-меры: {best_f1}, при параметрах:')
print(f'Лучшее количество итераций: {best_iter};')
print(f'Лучший метод: {best_solver}.')

Максимальное достигнутое значение f1-меры: 0.47193877551020413, при параметрах:
Лучшее количество итераций: 25;
Лучший метод: lbfgs.


#### Вывод

С помощью увеличения размера выборки удалось добиться улучшения показателя f1-меры:
- Дерева Решений до 0.59;
- Случайного Леса до 0.61. 

Модель Логистической Регрессии не показала улучшения.

### Уменьшение выборки

Для увеличения выборки создам функциию, разделяющую выборку в соответствии со значением целевого признака, уменьшающую количество строк соответствующих негативному целевому признаку и меняющую порядок строк результата.

In [87]:
def downsample(features, target, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]
    
    target_zeros = target[target == 0]
    target_ones = target[target == 1]

    features_downsampled = pd.concat(
        [features_zeros.sample(frac=fraction, random_state=568702396)] + [features_ones])
    target_downsampled = pd.concat(
        [target_zeros.sample(frac=fraction, random_state=568702396)] + [target_ones])
    
    features_downsampled, target_downsampled = shuffle(features_downsampled, target_downsampled, random_state=568702396)
    
    return features_downsampled, target_downsampled

features_downsampled, target_downsampled = downsample(features_train, target_train, 0.5)

In [88]:
# сравню размеры
display(features_downsampled.shape)
target_downsampled.shape

(3617, 11)

(3617,)

In [89]:
# проверю баланс классов
print('Количество классов в целевом признаке: \n')
print(target_downsampled.value_counts())

print('\n ////////////////////////// \n')

print('В процентном соотношении:')
target_downsampled.value_counts().div(
    target_downsampled.value_counts().sum()
).mul(100).round(2)

Количество классов в целевом признаке: 

0    2383
1    1234
Name: exited, dtype: int64

 ////////////////////////// 

В процентном соотношении:


0    65.88
1    34.12
Name: exited, dtype: float64

#### Дерево решений на уменьшенной выборке

In [119]:
best_criteria = 0
best_depth = 0
best_split = 0
best_leaf = 0
best_f1 = 0

criteria = ['gini', 'entropy']

for i in range(len(criteria)):
    for depth in range(2, 6):
        for split in range(2, 6):
            for leaf in range(2, 6):
                model_tree_down = DecisionTreeClassifier(
                    random_state=568702396,
                    criterion=criteria[i],
                    max_depth=depth,
                    min_samples_split=split,
                    min_samples_leaf=leaf)
                
                model_tree_down.fit(features_downsampled, target_downsampled)
                predictions_valid = model_tree_down.predict(features_valid)
                
                f1 = f1_score(target_valid, predictions_valid)
                if f1 > best_f1:
                    best_f1 = f1
                    best_criteria = criteria[i]
                    best_depth = depth
                    best_split = split
                    best_leaf = leaf
                    
                
print(f'Максимальное достигнутое значение f1-меры: {best_f1}, при параметрах:')
print(f'- критерион - {best_criteria};')
print(f'- глубина дерева решений = {best_depth};')
print(f'- минимальное количество примеров для разделения: {best_split};')
print(f'- минимальное количество выборок для листового узла: {best_leaf}.')

Максимальное достигнутое значение f1-меры: 0.5989304812834224, при параметрах:
- критерион - entropy;
- глубина дерева решений = 5;
- минимальное количество примеров для разделения: 2;
- минимальное количество выборок для листового узла: 3.


#### Случайный лес на уменьшенной выборке

In [120]:
best_criteria = 0
best_depth = 0
best_est = 0
best_leaf = 0
best_f1 = 0

criteria = ['gini', 'entropy']

for i in range(len(criteria)):
    for est in range(2, 53, 5):
        for depth in range(2, 6):
            for leaf in range(2, 6):
               model_forest_down = RandomForestClassifier(
                   random_state=568702396,
                   criterion=criteria[i],
                   max_depth=depth,
                   n_estimators=est,
                   min_samples_leaf=leaf) 
               
               model_forest_down.fit(features_downsampled, target_downsampled)
               predictions_valid = model_forest_down.predict(features_valid)
               
               f1 = f1_score(target_valid, predictions_valid)
               if f1 > best_f1:
                   best_f1 = f1
                   best_criteria = criteria[i]
                   best_depth = depth
                   best_est = est
                   best_leaf = leaf

print(f'Максимальное достигнутое значение f1-меры: {best_f1} при модели:')
print(f'- количество деревьев: {best_est};')
print(f'- критерион - {best_criteria};')
print(f'- глубина дерева решений: {best_depth};')
print(f'- минимальное количество выборок для листового узла: {best_leaf}.')

Максимальное достигнутое значение f1-меры: 0.620967741935484 при модели:
- количество деревьев: 42;
- критерион - gini;
- глубина дерева решений: 5;
- минимальное количество выборок для листового узла: 4.


#### Логистическая регрессия на уменьшенной выборке

In [131]:
best_iter = 0
best_f1 = 0
best_solver = 0

solv = ['newton-cg', 'lbfgs', 'liblinear', 'saga']

for i in range(len(solv)):
    for itera in [25, 50, 100, 200, 400, 800]:
        model_regression_down = LogisticRegression(random_state=568702396, solver=solv[i], max_iter=itera)
        
        model_regression_down.fit(features_downsampled, target_downsampled)
        predictions_valid = model_regression_down.predict(features_valid)
        
        f1 = f1_score(target_valid, predictions_valid)
        if f1 > best_f1:
            best_f1 = f1
            best_iter = itera
            best_solver = solv[i]
 
print(f'Максимальное достигнутое значение f1-меры: {best_f1}, при параметрах:')
print(f'Лучшее количество итераций: {best_iter};')
print(f'Лучший метод: {best_solver}.')

Максимальное достигнутое значение f1-меры: 0.4686299615877081, при параметрах:
Лучшее количество итераций: 25;
Лучший метод: newton-cg.


#### Вывод

С помощью уменьшения размера выборки удалось добиться улучшения показателя f1-меры:
- Дерева Решений до 0.59;
- Случайного Леса до 0.62. 

Модель Логистической Регрессии не показала улучшения.

### Комбинирование увеличения и уменьшения выборки

In [113]:
def combine(features, target, repeat, fraction):
    features_zeros = features[target == 0]
    features_ones = features[target == 1]

    target_zeros = target[target == 0]
    target_ones = target[target == 1]
    
    features_combined = pd.concat([features_zeros.sample(frac=fraction, random_state=568702396)] + [features_ones] * repeat)
    target_combined = pd.concat([target_zeros.sample(frac=fraction, random_state=568702396)] + [target_ones] * repeat)
        
    features_combined, target_combined = shuffle(features_combined, target_combined, random_state=568702396)

    return features_combined, target_combined

features_combined, target_combined = combine(features_train, target_train, 2, 0.75)

In [109]:
# сравню размеры
display(features_combined.shape)
target_combined.shape

(4851, 11)

(4851,)

In [110]:
# проверю баланс классов
print('Количество классов в целевом признаке: \n')
print(target_combined.value_counts())

print('\n ////////////////////////// \n')

print('В процентном соотношении:')
target_combined.value_counts().div(
    target_combined.value_counts().sum()
).mul(100).round(2)

Количество классов в целевом признаке: 

1    2468
0    2383
Name: exited, dtype: int64

 ////////////////////////// 

В процентном соотношении:


1    50.88
0    49.12
Name: exited, dtype: float64

#### Дерево решений на комбинированной выборке

In [121]:
best_criteria = 0
best_depth = 0
best_split = 0
best_leaf = 0
best_f1 = 0

criteria = ['gini', 'entropy']

for i in range(len(criteria)):
    for depth in range(2, 6):
        for split in range(2, 6):
            for leaf in range(2, 6):
                model_tree_combo = DecisionTreeClassifier(
                    random_state=568702396,
                    criterion=criteria[i],
                    max_depth=depth,
                    min_samples_split=split,
                    min_samples_leaf=leaf)
                
                model_tree_combo.fit(features_combined, target_combined)
                predictions_valid = model_tree_combo.predict(features_valid)
                
                f1 = f1_score(target_valid, predictions_valid)
                if f1 > best_f1:
                    best_f1 = f1
                    best_criteria = criteria[i]
                    best_depth = depth
                    best_split = split
                    best_leaf = leaf
                    
                
print(f'Максимальное достигнутое значение f1-меры: {best_f1}, при параметрах:')
print(f'- критерион - {best_criteria};')
print(f'- глубина дерева решений = {best_depth};')
print(f'- минимальное количество примеров для разделения: {best_split};')
print(f'- минимальное количество выборок для листового узла: {best_leaf}.')

Максимальное достигнутое значение f1-меры: 0.5758513931888545, при параметрах:
- критерион - gini;
- глубина дерева решений = 5;
- минимальное количество примеров для разделения: 2;
- минимальное количество выборок для листового узла: 5.


#### Случайный лес на комбинированной выборке

In [122]:
best_criteria = 0
best_depth = 0
best_est = 0
best_leaf = 0
best_f1 = 0

criteria = ['gini', 'entropy']

for i in range(len(criteria)):
    for est in range(2, 53, 5):
        for depth in range(2, 6):
            for leaf in range(2, 6):
               model_forest_combo = RandomForestClassifier(
                   random_state=568702396,
                   criterion=criteria[i],
                   max_depth=depth,
                   n_estimators=est,
                   min_samples_leaf=leaf) 
               
               model_forest_combo.fit(features_combined, target_combined)
               predictions_valid = model_forest_combo.predict(features_valid)
               
               f1 = f1_score(target_valid, predictions_valid)
               if f1 > best_f1:
                   best_f1 = f1
                   best_criteria = criteria[i]
                   best_depth = depth
                   best_est = est
                   best_leaf = leaf

print(f'Максимальное достигнутое значение f1-меры: {best_f1} при модели:')
print(f'- количество деревьев: {best_est};')
print(f'- критерион - {best_criteria};')
print(f'- глубина дерева решений: {best_depth};')
print(f'- минимальное количество выборок для листового узла: {best_leaf}.')

Максимальное достигнутое значение f1-меры: 0.6157575757575757 при модели:
- количество деревьев: 32;
- критерион - gini;
- глубина дерева решений: 5;
- минимальное количество выборок для листового узла: 2.


#### Логистическая регрессия на комбинированной выборке

In [123]:
best_iter = 0
best_f1 = 0
best_solver = 0

solv = ['newton-cg', 'lbfgs', 'liblinear', 'saga']

for i in range(len(solv)):
    for itera in [25, 50, 100, 200, 400, 800]:
        model_regression_combo = LogisticRegression(random_state=568702396, solver=solv[i], max_iter=itera)
        
        model_regression_combo.fit(features_combined, target_combined)
        predictions_valid = model_regression_combo.predict(features_valid)
        
        f1 = f1_score(target_valid, predictions_valid)
        if f1 > best_f1:
            best_f1 = f1
            best_iter = itera
            best_solver = solv[i]
 
print(f'Максимальное достигнутое значение f1-меры: {best_f1}, при параметрах:')
print(f'Лучшее количество итераций: {best_iter};')
print(f'Лучший метод: {best_solver}.')

Максимальное достигнутое значение f1-меры: 0.4851157662624035, при параметрах:
Лучшее количество итераций: 25;
Лучший метод: newton-cg.


#### Вывод

С помощью комбинирования методов уменьшения и увеличения выборки выборки удалось добиться улучшения показателя f1-меры:
- Случайного Леса до 0.61. 

Модели Дерева Решений и Логистической Регрессии не показали значительного улучшения.

### Вывод

Науличшие показатели f1-меры были достигнуты при метода Случайного леса. 

Для определения наилучшей модели сравню показатели f1-меры и AUC-ROC трех обученных моделей Случайного леса, используя тестовую выборку.

## Тестирование модели

### Случайный лес, модель увеличенной выборки

In [153]:
predictions_test = model_forest_up.predict(features_test)
               
f1_score(target_test, predictions_test)

0.60431654676259

Сравню полученное значение f1-меры с показателем AUC-ROC.

In [159]:
probabilities_test = model_forest_up.predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]

auc_roc = roc_auc_score(target_test, probabilities_one_test)

auc_roc

0.847274855749432

### Случайный лес, модель уменьшенной выборки

In [160]:
predictions_valid = model_forest_down.predict(features_test)
               
f1_score(target_test, predictions_valid)

0.6091794158553546

Сравню полученное значение f1-меры с показателем AUC-ROC.

In [161]:
probabilities_test = model_forest_down.predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]

auc_roc = roc_auc_score(target_test, probabilities_one_test)

auc_roc

0.8470882284441607

### Случайный лес, модель комбинированной выборки

In [162]:
predictions_valid = model_forest_combo.predict(features_test)
        
f1_score(target_test, predictions_valid)

0.6142322097378277

Сравню полученное значение f1-меры с показателем AUC-ROC.

In [163]:
probabilities_test = model_forest_combo.predict_proba(features_test)
probabilities_one_test = probabilities_test[:, 1]

auc_roc = roc_auc_score(target_test, probabilities_one_test)

auc_roc

0.845393158952481

### Проверка на адекватность

С целью проверки на адекватность использую "глупую модель" DummyClassifier.

In [164]:
from sklearn.model_selection import GridSearchCV
from sklearn.dummy import DummyClassifier

dummy = DummyClassifier(random_state=568702396)

params_dummy = [{
    'strategy': ['most_frequent', 'prior', 'stratified', 'uniform', 'constant'],
    'constant': [0, 1]
}]

model_dummy = GridSearchCV(dummy,
                      param_grid=params_dummy,
                      scoring='f1')

model_dummy.fit(features_combined, target_combined)
model_dummy.best_params_

{'constant': 1, 'strategy': 'constant'}

In [166]:
predictions_dummy = model_dummy.predict(features_test)

display(f'f1-мера: {f1_score(target_test, predictions_dummy)}')

'f1-мера: 0.3381803074366431'

### Вывод

Было дано задание построить модель, способную на основе данных клиента спрогнозировать расторгнет ли клиент в скором времени договор с банком.  
Для выполнения проекта были предоставлены исторические данные банка.  
Была поставлена задача предоставить модель с наибольшим достигнутым значением f1-меры, но не меньше 0.59.

По итогам исследования достигнуто значение f1-меры 0.61 на тестовой выборке, что соответствует требованиям заказчика.  
Значение достигнуто путем использования метода Случайного леса и кобминированного метода увеличения и уменьшения выборки для борьбы с дисбалансом - модель "model_forest_combo". 

Модели обученные отдельно на увеличенной и уменьшенной выборках показали меньшее значение f1-меры на тестовой выборке.  
Значения AUC-ROC между тремя финальными моделями не показали значимых различий.